File: Backend/models/conversation.py
from sqlalchemy import BigInteger, CheckConstraint, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from database import Base
from models.audit import AuditMixin


sqlite_compatible_bigint = BigInteger().with_variant(Integer, "sqlite")


class Conversation(AuditMixin, Base):
    __tablename__ = "conversations"

    id = Column(sqlite_compatible_bigint, primary_key=True, autoincrement=True)
    user_id = Column(
        BigInteger,
        ForeignKey("users.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    title = Column(String(256), nullable=False, default="New conversation")

    user = relationship("User", back_populates="conversations")
    messages = relationship(
        "Message",
        back_populates="conversation",
        cascade="all, delete-orphan",
        order_by=lambda: (Message.created_at, Message.id),
    )


class Message(Base):
    __tablename__ = "messages"
    __table_args__ = (
        CheckConstraint(
            "role IN ('user', 'assistant')",
            name="ck_messages_role",
        ),
    )

    id = Column(sqlite_compatible_bigint, primary_key=True, autoincrement=True)
    conversation_id = Column(
        BigInteger,
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    role = Column(String(16), nullable=False)
    content = Column(Text, nullable=False)
    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    conversation = relationship("Conversation", back_populates="messages")

File: Frontend/types/conversation.ts
export type Conversation = {
  id: number;
  title: string;
  created_at: string;
  updated_at: string;
};

export type CreatedConversation = {
  conversation_id: number;
  title: string;
  created_at: string;
};

export type ConversationMessage = {
  id: number;
  conversation_id: number;
  role: "user" | "assistant";
  content: string;
  created_at: string;
};

export type MessageExchange = {
  conversation_id: number;
  conversation_title: string;
  user_message: ConversationMessage;
  assistant_message: ConversationMessage;
};
